crates/pallet-erc20-support/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+
+/// Utility alias for easy access to the [`Config::AccountId`].
+type AccountIdOf<T, I> = <T as Config<I>>::AccountId;
+
+/// Utility alias for easy access to the [`Currency::Balance`] of the [`Config::Currency`] type.
+type BalanceOf<T, I> = <<T as Config<I>>::Currency as Currency<AccountIdOf<T, I>>>::Balance;
+
+// We have to temporarily allow some clippy lints. Later on we'll send patches to substrate to
+// fix them at their end.
+#[allow(clippy::missing_docs_in_private_items)]
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::{pallet_prelude::*, sp_runtime::traits::MaybeDisplay, sp_std::fmt::Debug};
+
+    use super::*;
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T, I = ()>(_);
+
+    /// Configuration trait of this pallet.
+    #[pallet::config]
+    pub trait Config<I: 'static = ()>: frame_system::Config {
+        /// The user account identifier type.
+        type AccountId: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Debug
+            + MaybeDisplay
+            + Ord
+            + MaxEncodedLen;
+
+        /// The currency to be exposed as ERC20 token.
+        type Currency: Currency<AccountIdOf<Self, I>>;
+
+        /// Interface into ERC20 metadata implementation.
+        type Metadata: Metadata;
+    }
+
+    /// ERC20-style approvals data.
+    /// (Owner => Allowed => Amount).
+    #[pallet::storage]
+    #[pallet::getter(fn approvals)]
+    pub type Approvals<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        AccountIdOf<T, I>,
+        Blake2_128Concat,
+        AccountIdOf<T, I>,
+        BalanceOf<T, I>,
+    >;
+
+    /// Possible errors.
+    #[pallet::error]
+    pub enum Error<T, I = ()> {
+        /// Spender not allowed to transfer tokens on behalf of owner.
+        SpenderNotAllowed,
+        /// Spender can't transfer tokens more than allowed.
+        SpendMoreThanAllowed,
+    }
+}
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+    /// Returns the amount of tokens in existence.
+    pub fn total_supply() -> BalanceOf<T, I> {
+        T::Currency::total_issuance()
+    }
+
+    /// Returns the amount of tokens owned by provided account.
+    pub fn balance_of(owner: &AccountIdOf<T, I>) -> BalanceOf<T, I> {
+        T::Currency::total_balance(owner)
+    }
+
+    /// Returns the remaining number of tokens that spender will be allowed to spend on behalf of
+    /// owner. This is zero by default.
+    pub fn allowance(owner: &AccountIdOf<T, I>, spender: &AccountIdOf<T, I>) -> BalanceOf<T, I> {
+        Self::approvals(owner, spender).unwrap_or_default()
+    }
+
+    /// Sets amount as the allowance of spender over the caller’s tokens.
+    pub fn approve(owner: AccountIdOf<T, I>, spender: AccountIdOf<T, I>, amount: BalanceOf<T, I>) {
+        <Approvals<T, I>>::insert(owner, spender, amount);
```

Comment:
Should we prohibit approvals where `owner == spender`?

Reply:
I guess we can prohibit it but in general if we check openzeppelin ERC20 implementation - https://github.com/OpenZeppelin/openzeppelin-contracts/blob/618304cc01e718cbdd87059ce2978b748c15050f/contracts/token/ERC20/ERC20.sol#L283. This case is not prohibited.

Reply:
But then they also don't have the aspect that the owner doesn't consume allowance, like we have: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/618304cc01e718cbdd87059ce2978b748c15050f/contracts/token/ERC20/ERC20.sol#L311

There is owner uses `transferFrom` but doesn't create an allowance for themselves they'll get a rever just like anyone else would.
Out implementation, on the other hand, has the `If caller is "sender", it can spend as much as it wants.` commend and the associated logic, which doesn't seem right. I think we should rather make it go through the usual allowance restrictions, even for the owner.

Reply:
Got it, added the restriction logic for the same owner and spender - https://github.com/humanode-network/humanode/pull/746/commits/78e8bf8fb04bd40300a160ac3d84679c57ca21fa.

Reply:
You have to pick one, not both. :D

Either we make allowances limit the spending for the owner (via the `transferFrom` calls only!) - in which case the owner can add them for themselves, or we do not allow the allowances to limit owners' spending when using `transferFrom` - then we must not allow adding the allowances for the owner in the first place. I didn't look into the EIPs or anything, but this is what just makes sense logically. I suggest checking the ERC20 spec - maybe it explicitly specifies what has to happen, so we don't have to choose/guess here.